Add regression test for n-link pendulum model

oliverlee opened the pull request:
This commit resolves https://github.com/pydy/pydy/issues/108.

oliverlee committed:
Add regression test for n-link pendulum model

This commit resolves https://github.com/pydy/pydy/issues/108.

oliverlee commented:
Regression test is failing, likely due to tolerance errors near zero. We currently test the integration output using ```numpy.allclose``` with ```rtol=1e-7, atol=0```.

@moorepants Do you have a suggestion on absolute tolerance?

moorepants commented:
The atol is needed to compare values to zero and the rtol for comparing to any other values.

We may need an atol value if comparisons to zero are happening.

oliverlee committed:
Set absolute tolerance for n-link pendulum test

oliverlee commented:
@moorepants The test still fails after adding absolute tolerance 1e-7.
Can you try this on your machine? And create the pendulum data file by uncommenting this line: https://github.com/oliverlee/pydy/blob/8ebad6af6cda935fac7265cc577f426be844b44d/pydy/tests/test_models.py#L122

This passes on my machine and I don't know how (or if I can) create the data on a Travis machine.

moorepants commented:
Yeah, this is tricky. I'll have to look at it tomorrow. Too much pydy already today.

oliverlee commented:
I don't know if you guys were aware, but there's a pretty large difference when using the different integrator methods

![benchmark-states](https://cloud.githubusercontent.com/assets/2965031/10576672/51678a72-7665-11e5-8839-bacad4bb7e78.png)

![benchmark-error](https://cloud.githubusercontent.com/assets/2965031/10576675/55423cb4-7665-11e5-9713-9ee5d5f7da03.png)

```
for state 0 simulation exceeds 1e10*eps difference between indices: [421, 999]
for state 1 simulation exceeds 1e10*eps difference between indices: [419, 999]
for state 2 simulation exceeds 1e10*eps difference between indices: [456, 999]
for state 3 simulation exceeds 1e10*eps difference between indices: [505, 982]
for state 4 simulation exceeds 1e10*eps difference between indices: [203, 999]
for state 5 simulation exceeds 1e10*eps difference between indices: [201, 991]
for state 6 simulation exceeds 1e10*eps difference between indices: [385, 999]
for state 7 simulation exceeds 1e10*eps difference between indices: [375, 962]
1e10*eps = 2.220446049250313e-06
```

oliverlee commented:
Okay after changing the integration settings (reducing time step and reducing total simulation time) , it seems to have a much closer match:

![benchmark-error](https://cloud.githubusercontent.com/assets/2965031/10579682/cbf4338a-7679-11e5-8cc6-75a51c418c6a.png)

![benchmark-states](https://cloud.githubusercontent.com/assets/2965031/10579683/cf8fe8c2-7679-11e5-81be-6442c7553d91.png)

so hopefully this works.

oliverlee committed:
Change simulation parameters and test tolerances

Reduce total simulation time and decrease simulation time step so that
we do not compare states after the system trajectory has diverged due to
chaotic behavior. Allow larger tolerances in comparison of system
generalized speeds.

oliverlee commented:
If you need to create the same plots, you can use this: https://gist.github.com/oliverlee/5a9d3e44d57313982e4b

oliverlee commented:
This still fails. For Python 2.7 oldest:
```
  File "/home/travis/build/pydy/pydy/pydy/tests/test_models.py", line 126, in test_n_link_pendulum_on_cart_regression
    np.testing.assert_allclose(x[:, :n+1], expected_x[:, :n+1], rtol, atol)
  File "/home/travis/miniconda/envs/test-env/lib/python2.7/site-packages/numpy/testing/utils.py", line 1347, in assert_allclose
    verbose=verbose, header=header)
  File "/home/travis/miniconda/envs/test-env/lib/python2.7/site-packages/numpy/testing/utils.py", line 708, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-07, atol=1e-07
(mismatch 0.00749999999999%)
```

moorepants commented:
This may be tough to get working. Maybe reducing the integration to something very simple would be best. The three backends that evaluate the eom's could certainly output numerically different answers. Tracking down the numerical inconsistency will be hard. I don't have any great suggestions right now for this issue.

moorepants commented:
This is probably worth rebasing/merging with master and running tests again.

oliverlee commented:
Works on my end after rebasing off master.

moorepants commented:
Testing the integration for so many steps isn't that necessary. Checking the rhs eval is sufficient.
